pkg/graveler/committed/merge.go

```diff
@@ -0,0 +1,425 @@
+func (m *merger) applyDestRangeSourceKey(destRange *Range, sourceValue *graveler.ValueRecord) error {
+	if bytes.Compare(destRange.MinKey, sourceValue.Key) > 0 { // source before dest range
+		baseValue, err := m.baseKeyGE(sourceValue.Key)
+		if err != nil {
+			return fmt.Errorf("base key GE: %w", err)
+		}
+		if baseValue != nil && bytes.Equal(sourceValue.Identity, baseValue.Identity) { // dest deleted this record
+			m.haveSource = m.source.Next()
+		} else {
+			if baseValue != nil && bytes.Equal(sourceValue.Key, baseValue.Key) { // conflict
+				return graveler.ErrConflictFound
+			}
+			// source added this record
+			err := m.writeRecord(sourceValue)
+			if err != nil {
+				return fmt.Errorf("write source record: %w", err)
+			}
+			m.haveSource = m.source.Next()
+		}
+		return nil
+	}
+
+	if bytes.Compare(destRange.MaxKey, sourceValue.Key) < 0 { // dest range before source
+		baseRange, err := m.baseRangeGE(graveler.Key(destRange.MinKey))
+		if err != nil {
+			return fmt.Errorf("base range GE: %w", err)
+		}
+		if baseRange != nil && destRange.ID == baseRange.ID { // source deleted this range
+			m.haveDest = m.dest.NextRange()
+			return nil
+		} else {
+			return graveler.ErrConflictFound // conflict
+		}
+	}
+	// dest is at start of range which we need to scan, enter it
+	m.haveDest = m.dest.Next()
+	return nil
+}
+
+// applySourceRangeDestKey applies merging when dest inside range and source on range header
+func (m *merger) applySourceRangeDestKey(sourceRange *Range, destValue *graveler.ValueRecord) error {
+	if bytes.Compare(sourceRange.MinKey, destValue.Key) > 0 { // dest before source range
+		baseValue, err := m.baseKeyGE(destValue.Key)
+		if err != nil {
+			return fmt.Errorf("base key GE: %w", err)
+		}
+		if baseValue != nil && bytes.Equal(destValue.Identity, baseValue.Identity) { // source deleted this record
+			m.haveSource = m.source.Next()
+		} else {
+			if baseValue != nil && bytes.Equal(destValue.Key, baseValue.Key) { // conflict
+				return graveler.ErrConflictFound
+			}
+			// dest added this record
+			err := m.writeRecord(destValue)
+			if err != nil {
+				return fmt.Errorf("write dest record: %w", err)
+			}
+			m.haveDest = m.dest.Next()
+		}
+		return nil
+	}
+
+	if bytes.Compare(sourceRange.MaxKey, destValue.Key) < 0 { // source range before dest
+		baseRange, err := m.baseRangeGE(graveler.Key(sourceRange.MinKey))
+		if err != nil {
+			return fmt.Errorf("base range GE: %w", err)
+		}
+		if baseRange != nil && sourceRange.ID == baseRange.ID { // dest deleted this range
+			m.haveSource = m.source.NextRange()
+			return nil
+		} else {
+			return graveler.ErrConflictFound // conflict
+		}
+	}
+	// source is at start of range which we need to scan, enter it
+	m.haveSource = m.source.Next()
+	return nil
+}
+
+func (m *merger) merge() error {
+	m.haveSource, m.haveDest, _ = m.source.Next(), m.dest.Next(), m.base.Next()
```

Comment:
why are you calling base.next without checking the result, what happens on error?

Reply:
Done